packages/cli-kit/src/private/node/session/identity-token-validation.ts

```diff
@@ -1,40 +1,67 @@
 import {shopifyFetch} from '../../../public/node/http.js'
+import {cacheFetch} from '../conf-store.js'
 
-export async function validateIdentityToken(token: string) {
+class NotFoundError extends Error {
+  constructor(message: string) {
+    super(message)
+    this.name = 'NotFoundError'
+  }
+}
+
+export async function validateIdentityToken(token: string): Promise<boolean> {
   try {
-    const instrospectionURL = await getInstrospectionEndpoint()
-    const options = {
-      method: 'POST',
-      headers: {Authorization: `Bearer ${token}`, 'Content-Type': 'application/json'},
-      body: JSON.stringify({token}),
-    }
-    outputDebug(`Sending Identity Introspection request to URL: ${instrospectionURL}`)
+    return withIntrospectionURL<boolean>(async (introspectionURL: string) => {
+      const options = {
+        method: 'POST',
+        headers: {Authorization: `Bearer ${token}`, 'Content-Type': 'application/json'},
+        body: JSON.stringify({token}),
+      }
+      outputDebug(`Sending Identity Introspection request to URL: ${introspectionURL}`)
 
-    const response = await shopifyFetch(instrospectionURL, options)
+      const response = await shopifyFetch(introspectionURL, options)
 
-    if (response.ok && response.headers.get('content-type')?.includes('json')) {
-      // eslint-disable-next-line @typescript-eslint/no-explicit-any
-      const json: any = await response.json()
-      outputDebug(`The identity token is valid: ${json.valid}`)
-      return json.valid
-    } else {
-      const text = await response.text()
-      outputDebug(`The Introspection request failed with:
+      if (response.ok && response.headers.get('content-type')?.includes('json')) {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const json: any = await response.json()
+        outputDebug(`The identity token is valid: ${json.valid}`)
+        return json.valid
+      } else if (response.status === 404) {
+        throw new NotFoundError('Identity Introspection endpoint not found')
+      } else {
+        const text = await response.text()
+        outputDebug(`The Introspection request failed with:
  - status: ${response.status}
  - www-authenticate header: ${JSON.stringify(response.headers.get('www-authenticate'))}
  - body: ${JSON.stringify(text)}`)
-      return false
-    }
+        return false
+      }
+    })
     // eslint-disable-next-line no-catch-all/no-catch-all
   } catch (error) {
     outputDebug(`The identity token is invalid: ${error}`)
     return false
   }
 }
 
-async function getInstrospectionEndpoint(): Promise<string> {
+async function withIntrospectionURL<T>(fn: (introspectionUrl: string) => Promise<T>): Promise<T> {
+  const week = 7 * 24 * 60 * 60 * 1000
+  const introspectionURL = await cacheFetch('identity-introspection-url', getIntrospectionURL, week)
+  try {
+    return fn(introspectionURL)
+  } catch (error) {
+    if (error instanceof NotFoundError) {
```

Comment:
Note that the `catch` block assumes that `fn` will use the error `NotFoundError` type to communicate that the token was not found. While that's an obvious assumption at the time this PR is being written, it might not be obvious for someone looking at this code in months from now. I'd recommend using `Result` types here:

```js
const result = await fn(introspectionURL)
// Type the result.error instance such that you can do an exhaustive check on the various error types.
```